tiny_dnn/core/kernels/conv2d_op_opencl.h

```diff
@@ -46,22 +46,26 @@ class Conv2dOpenCLForwardOp : public core::OpKernel {
     CLCudaAPI::Context ctx   = context.device()->context();
     CLCudaAPI::Queue queue   = context.device()->queue();
 
+    size_t samples_num = in_data_t.shape()[0], out_size = out_data_t.shape()[1],
+           in_size = in_data_t.shape()[1];
     // TODO(edgar): check if we really need that
-    for (serial_size_t i = 0; i < in_data.size(); ++i) {
+    for (serial_size_t i = 0; i < samples_num; ++i) {
+      auto in_begin  = in_data_t.host_pointer(i, 0),
+           in_end    = in_data_t.host_pointer(i, 0) + in_size;
+      auto out_begin = out_data_t.host_pointer(i, 0),
+           out_end   = out_data_t.host_pointer(i, 0) + in_size;
+
       // Creates device buffers and copies the host data to these
       // device buffers.
+      auto dev_in = CLCudaAPI::Buffer<float_t>(ctx, queue, in_begin, in_end);
 
-      auto dev_in = CLCudaAPI::Buffer<float_t>(ctx, queue, in_data[i].begin(),
-                                               in_data[i].end());
-
-      auto dev_W =
-        CLCudaAPI::Buffer<float_t>(ctx, queue, W[0].begin(), W[0].end());
+      auto dev_W = CLCudaAPI::Buffer<float_t>(
+        ctx, queue, weights_t.host_pbegin(), weights_t.host_pend());
 
-      auto dev_bias =
-        CLCudaAPI::Buffer<float_t>(ctx, queue, bias[0].begin(), bias[0].end());
+      auto dev_bias = CLCudaAPI::Buffer<float_t>(
+        ctx, queue, bias_t.host_pbegin(), bias_t.host_pbegin());
```

Comment:
brief comment: we should start thinking about how are going to share memory between devices